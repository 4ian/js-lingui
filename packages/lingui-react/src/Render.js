// @flow
import React from 'react'

export type RenderProps = {
  render?: any,
  className?: string
}

type RenderComponentProps = {
  value: string | Array<any>
} & RenderProps

const Render = ({ render = 'span', className, value }: RenderComponentProps) => {
  // Built-in element: h1, p
  if (typeof render === 'string') {
    return React.createElement(render, { className }, value)
  }

  return React.isValidElement(render)
    // Custom element: <p className="lear' />
    ? React.cloneElement(render, {}, value)
    // Custom component: ({ translation }) => <a title={translation}>x</a>
    : React.createElement(render, { translation: value })
}

export default Render
