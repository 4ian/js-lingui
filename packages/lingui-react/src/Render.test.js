// @flow
import React from 'react'
import { shallow } from 'enzyme'
import Render from './Render'

describe('Render', function () {
  it('should render children wrapped in span', function () {
    const span = shallow(<Render>Just a text</Render>)
    expect(span).toMatchSnapshot()

    const withClass = shallow(<Render className="info">Just a text</Render>)
    expect(withClass).toMatchSnapshot()
  })

  it('should render custom element', function () {
    const builtin = shallow(<Render render="h1">Headline</Render>)
    expect(builtin).toMatchSnapshot()

    const element = shallow(<Render render={<h1 />}>Headline</Render>)
    expect(element).toMatchSnapshot()
  })

  it('should render custom component', function () {
    const component = shallow(<Render render={
      ({ translation }) => <a title={translation}>X</a>
    }>Title</Render>)
    expect(component.dive()).toMatchSnapshot()
  })
})
