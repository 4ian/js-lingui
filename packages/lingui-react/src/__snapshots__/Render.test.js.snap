// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render should render children wrapped in span 1`] = `
<span>
  Just a text
</span>
`;

exports[`Render should render children wrapped in span 2`] = `
<span
  className="info"
>
  Just a text
</span>
`;

exports[`Render should render custom component 1`] = `
<a
  title="Title"
>
  X
</a>
`;

exports[`Render should render custom element 1`] = `
<h1>
  Headline
</h1>
`;

exports[`Render should render custom element 2`] = `
<h1>
  Headline
</h1>
`;
